File: src/main/java/src/leetcode/linkedlist/l234_is_palindrome/L234.java
package leetcode.linkedlist.l234_is_palindrome;

import leetcode.linkedlist.ListNode;
import linkedlist.NodesKt;

public class L234 {
    public boolean isPalindrome (ListNode head) {
        if (head == null) return true;
        if (head.next == null) return true;

        ListNode slow = head;
        ListNode fast = head.next;
        ListNode flag = new ListNode(0);

        while (fast != null && fast.next != null) {
            fast = fast.next.next;

            ListNode slowNext = slow.next;

            slow.next = flag.next;
            flag.next = slow;
            slow = slowNext;
        }

        if (fast == null) {
            //odd
            return same(flag.next, slow.next);
        } else {
            //even
            ListNode slowNext = slow.next;
            slow.next = flag.next;
            flag.next = slow;
            return same(flag.next, slowNext);
        }
    }

    public boolean same (ListNode head1, ListNode head2) {
        ListNode p1 = head1;
        ListNode p2 = head2;
        while (p1 != null && p2 != null) {
            if (p1.val == p2.val) {
                p1 = p1.next;
                p2 = p2.next;
                continue;
            }
            return false;
        }

        if (p1 == null && p2 == null) {
            return true;
        } else {
            return false;
        }
    }

    public static void main (String[] args) {
        ListNode head1 = NodesKt.arrayToList(1, 2, 3, 4, 5);
        ListNode head2 = NodesKt.arrayToList(1,2,3,4,5,4,3,2,1);
        ListNode head3 = NodesKt.arrayToList(1,2,3,4,5,4,3,2,1);
        ListNode head4 = NodesKt.arrayToList(1,2,1);
        ListNode head5 = NodesKt.arrayToList(1,2);
        ListNode head6 = NodesKt.arrayToList(1);
        ListNode head7 = NodesKt.arrayToList();

        L234 obj = new L234();
        System.out.println(obj.isPalindrome(head1));
        System.out.println(obj.isPalindrome(head2));
        System.out.println(obj.isPalindrome(head3));
        System.out.println(obj.isPalindrome(head4));
        System.out.println(obj.isPalindrome(head5));
        System.out.println(obj.isPalindrome(head6));
        System.out.println(obj.isPalindrome(head7));
    }
}

File: src/main/java/src/leetcode/linkedlist/l234_is_palindrome/L234.kt
package linkedlist

import common.ListNode
import common.arrayToList

fun isPalindrome(head: ListNode?): Boolean {
    head ?: return true
    head.next ?: return true

    val flag = ListNode(-1000)

    var fast = head
    var slow = head
    var slowNext: ListNode?

    while (fast?.next?.next != null) {
        fast = fast.next.next
        slowNext = slow!!.next

        slow.next = flag.next
        flag.next = slow
        slow = slowNext
    }

    return if (fast!!.next == null) {
        //odd len
        theSameLinkedList(slow!!.next, flag.next)
    } else {
        //even len
        slowNext = slow!!.next
        slow.next = flag.next
        flag.next = slow

        theSameLinkedList(slowNext, flag.next)
    }
}

fun theSameLinkedList(first: ListNode, second: ListNode): Boolean {
    var f: ListNode? = first
    var s: ListNode? = second
    while (f != null && s != null) {
        if (f.value == s.value) {
            f = f.next
            s = s.next
            continue
        }
        return false
    }
    return true
}

fun main() {
    println(isPalindrome(arrayToList(1, 2)))
    println(isPalindrome(arrayToList(1, 2, 1)))
    println(isPalindrome(arrayToList(1, 2, 2, 1)))
    println(isPalindrome(arrayToList(1, 2, 3, 3, 2, 1)))
    println(isPalindrome(arrayToList(1)))
    println(isPalindrome(arrayToList()))
}